Update to 24.2 WIP

anthony-sarkis opened the pull request:
Related to https://github.com/tairov/llama2.mojo/issues/88


Work in progress, still getting these errors. I'm new to Mojo so if anyone can assist:

```
/root/llama2.mojo/llama2.mojo:506:20: error: 'String' value has no attribute 'bitcast'
    buf.data = data.bitcast[DType.uint8]()
               ~~~~^~~~~~~~
/root/llama2.mojo/llama2.mojo:517:29: error: no matching method in call to 'load'
        a.load[width=_nelts](j, a.load[width=_nelts](j) + b.load[width=_nelts](j))
        ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: method argument #1 cannot be converted from 'SIMD[f32, _nelts]' to 'Int'
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 1 positional argument, got 2
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:543:39: error: no matching method in call to 'load'
        o.offset(j).load[width=_nelts](0, val)
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 1 positional argument, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:580:29: error: no matching method in call to 'load'
        x.load[width=_nelts](start + ii, val)
        ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: method argument #1 cannot be converted from 'SIMD[f32, _nelts]' to 'Int'
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 1 positional argument, got 2
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:589:29: error: no matching method in call to 'load'
        x.load[width=_nelts](start + ii, x.load[width=_nelts](start + ii) / ssum)
        ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: method argument #1 cannot be converted from 'SIMD[f32, _nelts]' to 'Int'
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 1 positional argument, got 2
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:821:48: error: no matching method in call to 'load'
                    state.xb.load[width=_nelts](xb_offset + i, xbi)
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: method argument #1 cannot be converted from 'SIMD[f32, _nelts]' to 'Int'
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 1 positional argument, got 2
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:850:40: error: no matching method in call to 'load'
            state.hb.load[width=_nelts](i, hbi * state.hb2.load[width=_nelts](i))
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: method argument #1 cannot be converted from 'SIMD[f32, _nelts]' to 'Int'
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 1 positional argument, got 2
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:881:32: error: invalid call to 'rand': missing 1 required positional argument: 'size'
    var r = rand[DType.float32](1)
            ~~~~~~~~~~~~~~~~~~~^~~
/root/llama2.mojo/llama2.mojo:1:1: note: function declared here
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:50:37: warning: 'SIMD[T, _width]' value is unused
        self.data.load[width=_width](newVal)
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
        _ = 
/root/llama2.mojo/llama2.mojo:111:38: error: no matching method in call to 'load'
        return self._data.load[nelts](idx)
               ~~~~~~~~~~~~~~~~~~~~~~^~~~~
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 1 positional argument, got 2
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: callee parameter #2 has 'Intable' type, but value has type 'Int'
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:119:38: error: no matching method in call to 'load'
        return self.load[width=nelts](indices[0] * self._shape[1] + indices[1])
               ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/root/llama2.mojo/llama2.mojo:110:8: note: candidate not viable: unknown keyword parameter: 'width'
    fn load[nelts: Int](self, idx: Int) -> SIMD[DType.float32, nelts]:
       ^
/root/llama2.mojo/llama2.mojo:113:8: note: candidate not viable: unknown keyword parameter: 'width'
    fn load[nelts: Int](self, *indices: Int) -> SIMD[DType.float32, nelts]:
       ^
/root/llama2.mojo/llama2.mojo:121:8: note: candidate not viable: unknown keyword parameter: 'width'
    fn load[
       ^
/root/llama2.mojo/llama2.mojo:129:8: note: candidate not viable: missing 1 required positional argument: 'val'
    fn load[nelts: Int](self, idx: Int, val: SIMD[DType.float32, nelts]):
       ^
/root/llama2.mojo/llama2.mojo:130:38: error: no matching method in call to 'load'
        return self._data.load[nelts](idx, val)
               ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 1 positional argument, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:1:1: note: candidate not viable: expected at most 2 positional arguments, got 3
from algorithm import sum
^
/root/llama2.mojo/llama2.mojo:133:34: error: no matching method in call to 'load'
        return self.load[width=1](idx, val)
               ~~~~~~~~~~~~~~~~~~^~~~~~~~~~
/root/llama2.mojo/llama2.mojo:110:8: note: candidate not viable: expected at most 2 positional arguments, got 3
    fn load[nelts: Int](self, idx: Int) -> SIMD[DType.float32, nelts]:
       ^
/root/llama2.mojo/llama2.mojo:113:8: note: candidate not viable: unknown keyword parameter: 'width'
    fn load[nelts: Int](self, *indices: Int) -> SIMD[DType.float32, nelts]:
       ^
/root/llama2.mojo/llama2.mojo:121:8: note: candidate not viable: expected at most 2 positional arguments, got 3
    fn load[
       ^
/root/llama2.mojo/llama2.mojo:129:8: note: candidate not viable: unknown keyword parameter: 'width'
    fn load[nelts: Int](self, idx: Int, val: SIMD[DType.float32, nelts]):
       ^
/root/llama2.mojo/llama2.mojo:382:40: error: 'List[SIMD[si8, 1]]' value has no attribute 'bitcast'
        var int32_ptr = config_data_raw.bitcast[DType.int32]()
                        ~~~~~~~~~~~~~~~^~~~~~~~
/root/llama2.mojo/llama2.mojo:469:27: error: 'List[SIMD[si8, 1]]' value has no attribute 'bitcast'
            var data = tmp.bitcast[DType.float32]()
                       ~~~^~~~~~~~
/root/.modular/pkg/packages.modular.com_max/bin/mojo: error: failed to parse the provided Mojo
```

mikowals commented on llama2.mojo:
```diff
@@ -379,7 +379,7 @@ struct Config:
         f.close()
         # correct Tensor type and shape for easy reading, without copying data
-        var int32_ptr = config_data_raw._steal_ptr().bitcast[DType.int32]()
+        var int32_ptr = config_data_raw.bitcast[DType.int32]()
```
Are you sure we don't still need to steal the pointer also?  `config_data_raw` is going to be deleted after last use which will free this pointer before we are done with it if it is still the owner.

anthony-sarkis replied:
I'm not sure

With it remaining the error is 

```
/home/anthony/llama2.mojo/llama2.mojo:382:40: error: 'List[SIMD[si8, 1]]' value has no attribute '_steal_ptr'
        var int32_ptr = config_data_raw._steal_ptr().bitcast[DType.int32]()
```

with it removed the error is
```
/home/anthony/llama2.mojo/llama2.mojo:382:40: error: 'List[SIMD[si8, 1]]' value has no attribute 'bitcast'
        var int32_ptr = config_data_raw.bitcast[DType.int32]()
```

So the theme seems to be `'List[SIMD[si8, 1]]' value has no attribute `

mikowals replied:
I am sure.  😃 The correct thing is to steal the pointer so we take ownership.  The method signature has just changed so it is a bit more elaborate now.  

The full correct line here is:
`var int32_ptr = DTypePointer[DType.int8](config_data_raw.steal_ptr().value).bitcast[DType.int32]()`. 

It can be split into multiple lines if you think it is clearer but the steps are steal the pointer, convert to DTypePointer, and then bitcast to the int32 representation we want.

mikowals commented on llama2.mojo:
```diff
@@ -466,7 +466,7 @@ struct TransformerWeights:
             var tmp = f.read_bytes(shape.num_elements() * sizeof[DType.float32]())
             bytes_read += shape.num_elements() * sizeof[DType.float32]()
-            var data = tmp._steal_ptr().bitcast[DType.float32]()
+            var data = tmp.bitcast[DType.float32]()
```
As above, we don't want `tmp` to keep ownership.

anthony-sarkis replied:
I am guessing that the bitcast errors may all be similar. I referenced the errors above with it either way

mikowals replied:
yes, same as above with changes to types is brackets and the variable name.

mikowals commented on llama2.mojo:
```diff
@@ -503,7 +503,7 @@ fn read_file(file_name: String, inout buf: FileBuf) raises:
     fd.close()
     buf.size = data._buffer.size
-    buf.data = data._steal_ptr().bitcast[DType.uint8]()
+    buf.data = data.bitcast[DType.uint8]()
```
As above, we don't want `data` to keep ownership.

mikowals reviewed (changes requested):
Thanks for undertaking this.  It is a fairly chunky set of changes for this release.

mikowals commented on llama2.mojo:
```diff
@@ -46,12 +46,12 @@ struct Accumulator[T: DType, width: Int]:
-        var newVal = self.data.simd_load[_width]() + val
-        self.data.simd_store[_width](newVal)
+        var newVal = self.data.load[width=_width]() + val
+        self.data.load[width=_width](newVal)
```
`self.data.store` I believe.

mikowals commented on llama2.mojo:
```diff
@@ -107,30 +107,30 @@ struct TensorSlice:
 
-    fn simd_store[nelts: Int](self, idx: Int, val: SIMD[DType.float32, nelts]):
-        return self._data.simd_store[nelts](idx, val)
+    fn load[nelts: Int](self, idx: Int, val: SIMD[DType.float32, nelts]):
```
still need to be `store` and change `nelts` -> `width`

mikowals commented on llama2.mojo:
```diff
@@ -107,30 +107,30 @@ struct TensorSlice:
 
     fn __setitem__(self, idx: Int, val: SIMD[DType.float32, 1]):
-        return self.simd_store[1](idx, val)
+        return self.load[width=1](idx, val)
```
`store` since we are setting the value.

mikowals commented on llama2.mojo:
```diff
@@ -815,10 +815,10 @@ fn transformer(
-                    ) + a * state.value_cache.simd_load[_nelts](v_offset + i)
-                    state.xb.simd_store[_nelts](xb_offset + i, xbi)
+                    ) + a * state.value_cache.load[width=_nelts](v_offset + i)
+                    state.xb.load[width=_nelts](xb_offset + i, xbi)
```
`store` since we are updating the value.

mikowals commented on llama2.mojo:
```diff
@@ -843,11 +843,11 @@ fn transformer(
             var hbi = initial_hb * (1.0 / (1.0 + math.exp(-initial_hb)))
             # Elementwise multiply with w3(x)
-            state.hb.simd_store[_nelts](i, hbi * state.hb2.simd_load[_nelts](i))
+            state.hb.load[width=_nelts](i, hbi * state.hb2.load[width=_nelts](i))
```
`store` since we are updating the value

mikowals commented on llama2.mojo:
```diff
@@ -514,7 +514,7 @@ fn accum(inout a: TensorF32, b: TensorF32) -> None:
     @parameter
     fn _acc[_nelts: Int](j: Int):
-        a.simd_store[_nelts](j, a.simd_load[_nelts](j) + b.simd_load[_nelts](j))
+        a.load[width=_nelts](j, a.load[width=_nelts](j) + b.load[width=_nelts](j))
```
this first instance of `a.load` needs to be `a.store`.  We are updating  the value at index `j`.

mikowals commented on llama2.mojo:
```diff
@@ -539,8 +539,8 @@ fn rmsnorm(
-        var val = weight.simd_load[_nelts](j) * ss * x.simd_load[_nelts](j)
-        o.offset(j).simd_store[_nelts](0, val)
+        var val = weight.load[width=_nelts](j) * ss * x.load[width=_nelts](j)
+        o.offset(j).load[width=_nelts](0, val)
```
`store` since we are updating the value.

mikowals commented on llama2.mojo:
```diff
@@ -576,8 +576,8 @@ fn softmax(inout x: TensorF32, start: Int, end: Int):
-        var val = math.exp(x.simd_load[_nelts](start + ii) - max_val)
-        x.simd_store[_nelts](start + ii, val)
+        var val = math.exp(x.load[width=_nelts](start + ii) - max_val)
+        x.load[width=_nelts](start + ii, val)
```
`store` since we are updating the value.

mikowals commented on llama2.mojo:
```diff
@@ -586,7 +586,7 @@ fn softmax(inout x: TensorF32, start: Int, end: Int):
     @parameter
     fn _norm[_nelts: Int](ii: Int):
-        x.simd_store[_nelts](start + ii, x.simd_load[_nelts](start + ii) / ssum)
+        x.load[width=_nelts](start + ii, x.load[width=_nelts](start + ii) / ssum)
```
`x.store` to replace first instance of `x.load`, since we are updating the value.

mikowals commented:
I pulled this and tried against Mojo 24.2 but I think maybe you are working against the nightly branch?  So these changes are likely to be needed in `24.4` assuming we get a release later this month that makes these new Mojo APIs public.

anthony-sarkis commented:
Hi Michael, thanks for all the updates 😎, I'll take a closer look and
check version tonight! 😊

On Mon, Apr 15, 2024, 7:58 PM Michael Kowalski ***@***.***>
wrote:

> I pulled this and tried against Mojo 24.2 but I think maybe you are
> working against the nightly branch? So these changes are likely to be
> needed in 24.4 assuming we get a release later this month that makes
> these new Mojo APIs public.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tairov/llama2.mojo/pull/89#issuecomment-2058142139>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEJ6DJAHEC5CLJLKT2XO3YDY5SHT7AVCNFSM6AAAAABGIKGSHOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANJYGE2DEMJTHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

anthony-sarkis commented:
Hello!

Okay so I think I fixed most of the items mentioned. Sorry about `simd_store` mix up I read the change log a little too fast and missed that.

I fixed a new error
https://github.com/anthony-sarkis/llama2.mojo/commit/017081e329c540770ee718f584f6cf36f5669e15

The only two sets of remaining errors seem to be
1. Attempting to add a None check to fix subscript error:
```
/home/anthony/llama2.mojo/llama2.mojo:886:13: error: 'None' is not subscriptable, it does not implement the `__getitem__`/`__setitem__` or `__refitem__` methods
        if r[0] < cdf:
```
https://github.com/anthony-sarkis/llama2.mojo/commit/394d31f8ad4056f78718943cffe99fd546b6a20a
But it still yields:
```
/home/anthony/llama2.mojo/llama2.mojo:886:12: error: 'None' does not implement the '__bool__' method
        if r:
```
Not surer if I'm missing something obvious but it seems like this isn't implemented in mojo the same as python?

2. The `bitcast` set of errors (which seems to actually be attribute errors

```
anthony@Anthony:~/llama2.mojo$ mojo llama2.mojo stories15M.bin -s 100 -n 256 -t 0.5 -i "Mojo is a language"
/home/anthony/llama2.mojo/llama2.mojo:506:20: error: 'String' value has no attribute 'bitcast'
    buf.data = data.bitcast[DType.uint8]()
               ~~~~^~~~~~~~
/home/anthony/llama2.mojo/llama2.mojo:886:12: error: 'None' does not implement the '__bool__' method
        if r:
           ^
/home/anthony/llama2.mojo/llama2.mojo:382:40: error: 'List[SIMD[si8, 1]]' value has no attribute '_steal_ptr'
        var int32_ptr = config_data_raw._steal_ptr().bitcast[DType.int32]()
                        ~~~~~~~~~~~~~~~^~~~~~~~~~~
/home/anthony/llama2.mojo/llama2.mojo:469:27: error: 'List[SIMD[si8, 1]]' value has no attribute 'bitcast'
            var data = tmp.bitcast[DType.float32]()
                       ~~~^~~~~~~~
mojo: error: failed to parse the provided Mojo
```

mikowals commented on llama2.mojo:
```diff
@@ -937,12 +939,12 @@ fn print_str(s: PointerString):
         var d1: Int = s[3].to_int()
         var d2: Int = s[4].to_int()
-        print_no_newline(chr(str2num(d1) * 16 + str2num(d2)))
+        print(chr(str2num(d1) * 16 + str2num(d2)))
```
Since `print_no_newline` was removed the correct way to mimic this now is with a second argument `end=""` otherwise the default behaviour adds the newline character.

mikowals commented on llama2.mojo:
```diff
@@ -937,12 +939,12 @@ fn print_str(s: PointerString):
     var p: Int = 0
     while s[p].to_int() != 0:
-        print_no_newline(chr(s[p].to_int()))
+        print(chr(s[p].to_int()))
```
as above, `end=""` needs to be a second argument provided to `print`

mikowals commented:
This error:
```console
/home/anthony/llama2.mojo/llama2.mojo:886:13: error: 'None' is not subscriptable, it does not implement the `__getitem__`/`__setitem__` or `__refitem__` methods
        if r[0] < cdf:
```
Is because `rand` that creates a random Tensor moved from the `random` module to the `tensor` module.  So `rand` needs to be added to the list of imports from `tensor` at the top of the file.  And the `from random` line can be removed.

tairov commented:
Hi @anthony-sarkis , I added a few fixes. Now it llama2 is compiling fine

anthony-sarkis commented:
Awesome, thanks everyone! 😊

On Wed, Apr 17, 2024, 1:44 AM Aydyn Tairov ***@***.***> wrote:

> Merged #89 <https://github.com/tairov/llama2.mojo/pull/89> into master.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tairov/llama2.mojo/pull/89#event-12504050837>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEJ6DJBD56NAAFTDUW25ANTY5YZAHAVCNFSM6AAAAABGIKGSHOVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJSGUYDIMBVGA4DGNY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
